### PR TITLE
Issue 44986/fix windows ui path

### PR DESCRIPTION
**Pull request by clippered**
#44968

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Can you squash this down to one commit -- and in particular remove the merge commits?

**Comment by nikomatsakis** [comment]
Also, I guess this doesn't really fix #44986 -- that seems to be unrelated?

**Comment by clippered** [comment]
Thanks for the feedback. I updated the PR to be a single commit. And I had a typo, all this time. It should refer to issue #44968

**Comment by estebank** [comment]
@clippered looks great, neat solution! Assuming travis passes, this is good to merge (r=me).

**Comment by kennytm** [comment]
@bors r- 

Basically every UI test failed on Windows, for details see https://ci.appveyor.com/project/rust-lang/rust/build/1.0.4949/job/fv87h5oarj17tgqr. Example diff:

```diff
 error[E0308]: mismatched types
-  --> $DIR/block-must-not-have-result-while.rs:13:9
+  --> C:/projects/rust/src/test/ui/block-result/block-must-not-have-result-while.rs:13:9
    |
 13 |         true //~  ERROR mismatched types
    |         ^^^^ expected (), found bool
    |
    = note: expected type `()`
               found type `bool`
 
 error: aborting due to previous error
```

**Comment by estebank** [comment]
@kennytm my bad, I should have done a bors try first.

@clippered could you change `src/runtest.rs`? What you'd have to do is the `output.replace(..., "$DIR")` for *both* `parent_dir.display().to_string().replace("\\", "\\\\")` and `parent_dir.display().to_string()`.

**Comment by clippered** [comment]
Apologies for that fix. I'll try to resolve this soon. Thanks for your help @estebank , but can you rephrase what you said? I somewhat did not get what you meant. Aren't the `parent_dir.display().to_string().replace("\\",` "\\\\")` and `parent_dir.display().to_string()` the same thing?

**Comment by estebank** [comment]
Sure thing! The compiler output to stderr can be one of two things (as far as we care about here):

 - Normal diagnostic output:
```
warning: unnecessary parentheses around assigned value
  --> /path/to/file/suggestions.rs:17:21
   |
17 |         let mut a = (1); // should suggest no `mut`, no parens
   |                     ^^^ help: remove these parentheses
   |
   = note: #[warn(unused_parens)] on by default
```
 - JSON output:
```json
{"message":"unnecessary parentheses around assigned value","code":null,"level":"warning","spans":[{"file_name":"/path/to/file/unused_parens_json_suggestion.rs","byte_start":976,"byte_end":989,"line_start":22,"line_end":22,"column_start":14,"column_end":27,"is_primary":true,"text":[{"text":"    let _a = (1 / (2 + 3));","highlight_start":14,"highlight_end":27}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"#[warn(unused_parens)] on by default","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"remove these parentheses","code":null,"level":"help","spans":[{"file_name":"/path/to/file/unused_parens_json_suggestion.rs","byte_start":976,"byte_end":989,"line_start":22,"line_end":22,"column_start":14,"column_end":27,"is_primary":true,"text":[{"text":"    let _a = (1 / (2 + 3));","highlight_start":14,"highlight_end":27}],"label":null,"suggested_replacement":"1 / (2 + 3)","expansion":null}],"children":[],"rendered":"    let _a = 1 / (2 + 3);"}],"rendered":null}
```

Under windows systems the json output looks the following way:
```json
{"message":"unnecessary parentheses around assigned value","code":null,"level":"warning","spans":[{"file_name":"C:\\projects\\rust\\src\\test\\ui\\lint\\unused_parens_json_suggestion.rs","byte_start":976,"byte_end":989,"line_start":22,"line_end":22,"column_start":14,"column_end":27,"is_primary":true,"text":[{"text":"    let _a = (1 / (2 + 3));","highlight_start":14,"highlight_end":27}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"#[warn(unused_parens)] on by default","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"remove these parentheses","code":null,"level":"help","spans":[{"file_name":"C:\\projects\\rust\\src\\test\\ui\\lint\\unused_parens_json_suggestion.rs","byte_start":976,"byte_end":989,"line_start":22,"line_end":22,"column_start":14,"column_end":27,"is_primary":true,"text":[{"text":"    let _a = (1 / (2 + 3));","highlight_start":14,"highlight_end":27}],"label":null,"suggested_replacement":"1 / (2 + 3)","expansion":null}],"children":[],"rendered":"    let _a = 1 / (2 + 3);"}],"rendered":null}
```

By replacing `\` with `\\` in `parent_dir_str` you fix the issue for the json output, but break it for the stderr output, as json needs the `\` to be escaped when printed out, but the terminal doesn't:

```
warning: unnecessary parentheses around assigned value
  --> c:\path\to\file\suggestions.rs:17:21
   |
17 |         let mut a = (1); // should suggest no `mut`, no parens
   |                     ^^^ help: remove these parentheses
   |
   = note: #[warn(unused_parens)] on by default
```

This doesn't happen on unix-like systems because the `/` doesn't need escaping on json strings. Because of all of this, in the normalization step you can get away with just doing a replacement of `c:\path\to\file\` and `c:\\path\\to\\file\\` for `$DIR/` and it would work for all cases. On unix-like system you'd be wasting time doing the replacement of `/path/to/file/` for `$DIR/` twice, as the second time nothing would be found and it would be a noop. Does this make sense?

**Comment by clippered** [comment]
I see, I will need to add the check for` --error-format json` before adding the replace then.

Pardon me for asking again but which test should I run to test the normal-diagnostic output? I have only ran `python x.py test src/test/ui`

**Comment by clippered** [comment]
@estebank I hope Travis accepts this one.
Thanks for your patience. I also realised that running python test will ignore all previous tests that passed and does not rerun them again, that's why I missed the other test cases.

**Comment by kennytm** [comment]
@estebank The try branch is not enabled on AppVeyor i.e. it won't be tested on Windows, so I don't think this can check whether the changeset is valid.

**Comment by estebank** [comment]
@kennytm wasn't aware of that... Ok, I'll try to get a windows machine to test this myself.

**Comment by aidanhs** [comment]
@estebank have you managed to find a windows machine to give this a try?

**Comment by estebank** [comment]
@aidanhs I have been on the road for the past week.

I'm giving it a @bors r+ for any errors to be caught by appveyor. **To anyone thinking about rolling this PR up, don't** so that if it fails (it shouldn't) it doesn't block the other PRs.

**Comment by kennytm** [comment]
@bors rollup-

The [rollup status](https://github.com/rust-lang/rust/pull/45080#issuecomment-335329295) has not cleared 😓

**Comment by estebank** [comment]
@kennytm, oops. Thank you for catching that.
